Added option to preserve original filter text

snyderp opened the pull request:
added a bool flag to `AdBlockClient.prototype.parse` to control whether this data should be preserved (since otherwise the text of ~ the entire filter list would be preserved in the uncommon case of wanting to use `findMatchingFilters`).

If `true`, then the return value of `findMatchingFilters` will have `matchingOrigRule` and `matchingExceptionOrigRule` properties that have the original filter rule text.  If `false`, then the `matchingOrigRule` and `matchingExceptionOrigRule` properties are `undefined`

bbondy commented on ad_block_client.cc:
```diff
@@ -294,6 +304,13 @@ void parseFilter(const char *input, const char *end, Filter *f,
               f->data[len - 1] = '\0';
               memcpy(f->data, input + i + 1, len - 1);
+
+              if (preserveRules == true) {
```
` if (preserveRules) {`

bbondy commented on filter.cc:
```diff
@@ -781,6 +798,15 @@ uint32_t Filter::serialize(char *buffer) {
   }
   totalSize += 1;
 
+  if (ruleDefinition) {
```
I don't think this is serializing correctly, I think we would need to store the `ruleDefinitionSize` in the above `snprintf`

bbondy replied:
Pls also add a test to `test/js/serializationTest.js` for that to ensure it works.

bbondy commented on filter.cc:
```diff
@@ -826,6 +852,14 @@ uint32_t Filter::deserialize(char *buffer, uint32_t bufferSize) {
   }
   consumed += hostLen + 1;
 
+  uint32_t ruleDefinitionLen = static_cast<uint32_t>(strlen(buffer + consumed));
```
Ditto I don't think this is deserializing correctly.

bbondy commented on ad_block_client.cc:
```diff
@@ -199,13 +200,16 @@ void parseFilter(const char *input, const char *end, Filter *f,
+    bool preserveRules) {
   FilterParseState parseState = FPStart;
   const char *p = input;
+  const char *q = p;
```
Let's call `q` `filterRuleEndPos` instead.

bbondy commented on ad_block_client.cc:
```diff
@@ -199,13 +200,16 @@ void parseFilter(const char *input, const char *end, Filter *f,
   memset(data, 0, sizeof data);
   int i = 0;
-
+  int ruleTextLength = 0;
```
I think we can just subtract `filterRuleEndPos` from the line start pos to get the same length. So we can probably get rid of this.

snyderp commented:
@bbondy rebased and change to address your comments. I also added a couple new sets of tests.

One thing that came up when writing this up though: i don't see anywhere that the `{Filter, CosmeticFilter}::{Serialize, Deserialize}` methods are called.  This ended up being very confusing, since I couldn't understand why the original filter rules we're being serialized (i missed that its actually done in `serializeFilters` in `ad_block_client.cc`).

Unless I'm missing where these methods are used, would you mind if I deleted all four methods, to avoid future possible confusion?

bbondy commented on ad_block_client.cc:
```diff
@@ -1420,6 +1451,14 @@ int serializeFilters(char * buffer, size_t bufferSizeAvail,
+      if (buffer) {
+        snprintf(buffer + bufferSize, bufferSizeAvail, "%s", f->ruleDefinition);
+      }
+      bufferSize += static_cast<int>(strlen(f->ruleDefinition));
```
any time serialization deserialization is changed, we need to update the adblock version, so I'm ok with this but you'd have to set it to version 5.

bbondy commented on filter.cc:
```diff
@@ -96,27 +103,41 @@ Filter::Filter(const Filter &other) {
+      size_t len = strlen(other.ruleDefinition) + 1;
+      ruleDefinition = new char[len];
+      snprintf(ruleDefinition, len, "%s", other.ruleDefinition);
+      ruleDefinitionLen = len - 1;
```
this is different from above, before the len included the 0 terminated character.
